### DISCUSSION TITLE
How to solve this problem

### DISCUSSION THREAD

**Discussion by 123xyl** [openmopac/mopac#139] [Q&A]
![image](https://user-images.githubusercontent.com/96686347/209274190-43f58fa9-ec65-46e0-95ea-a6bd5241eb68.png)
When testing, there are several output files like this, is this not running

**Comment by godotalgorithm** [comment]
That does look like a bug. The last atomic coordinate of the `Au_Add-H.mop` test input isn't being read correctly.

I assume from past discussions that you are still using the Flang compiler on Linux/ARM. Is this correct? Are you setting the environment variable `FORTRANOPT` to the value of `crlf` when you compile MOPAC? The CMake script should be requesting for you to do this. Flang has non-standard behavior with non-Linux endline characters in text files, which could be causing this problem. This environment variable changes how the Flang compiler parses endline characters.

If this doesn't fix your problem, then I will need to know more about it to help you. Can you give me a few more examples of tests that are failing and also some tests that are passing? Can you attach the full output files for these examples of tests that are failing?

I have a Raspberry Pi with an ARM processor in my office, and I can try to reproduce and fix this problem when I return to the office after my winter vacation.

**Reply by 123xyl** [reply]
how to setting the environment variable FORTRANOPT to the value of crlf ，

**Reply by 123xyl** [reply]
I am using the Flang compiler on Linux/ARM

**Reply by 123xyl** [reply]
![image](https://user-images.githubusercontent.com/96686347/209419698-46f2e883-3001-4366-bf87-f6d0c2f78166.png)
![image](https://user-images.githubusercontent.com/96686347/209419712-63805ec0-75eb-403b-98b1-48d6873d628f.png)
When running, it shows passed, but some examples output files like the above two pictures

**Reply by 123xyl** [reply]
![image](https://user-images.githubusercontent.com/96686347/209419825-5de0689c-f663-407b-8358-a9158b05c9ae.png)
The above output seems to be normal, but there is also a problem with the following image
![image](https://user-images.githubusercontent.com/96686347/209419886-9662235f-a186-4d65-94b9-6237809ebe4d.png)

**Comment by godotalgorithm** [comment]
The output of `Test_XENO.out` does suggest that I am right about your problem - the compiler is reading carriage returns as characters rather than ignoring them as part of end-line statements. Hopefully, this is the only problem. Otherwise, you have problems that I am unfamiliar with.

Setting environment variables depends on what your Linux shell is. If you are in a bash shell (the most common one), then the command is: `export FORTRANOPT=crlf`. If you are in a csh shell, then the command is `setenv FORTRANOPT crlf`. I think that one of these two commands should work. You need to run these commands before compiling MOPAC. I tried to put these commands in the CMake script, but CMake does not give easy access to environment variables at compile time.

**Reply by 123xyl** [reply]
![image](https://user-images.githubusercontent.com/96686347/209424537-40ab0e2c-dd32-4664-9b70-cc2581e2a3a6.png)
It still doesn't work, the result is the same as above

**Reply by 123xyl** [reply]
![image](https://user-images.githubusercontent.com/96686347/209425337-58480106-c132-444e-bfa0-3179eecc5aa1.png)
I set export FORTRANOPT=crlf before running and then a couple of tests failed

**Comment by 123xyl** [comment]
I've tested it a few times, and one or two use cases always fail each time, but on x86 they all pass   
![image](https://user-images.githubusercontent.com/96686347/209426983-a1dc2ab2-07cc-4db1-81b9-756905c89ec5.png)

**Comment by godotalgorithm** [comment]
It seems like `export FORTRANOPT=crlf` fixed many of your test failures. The failures that you are showing now look different each time. You also look like you are using MOPAC version 22.0.3, but the most recent version is now 22.0.6. There was a test dependency bug in 22.0.3 that was fixed in later versions. If you want to keep using 22.0.3, then you have to do serial testing to avoid this bug: use `ctest` instead of `ctest -j16`.

**Reply by 123xyl** [reply]
It does fix a lot of bugs, but why can't you use ctest -j16

**Reply by 123xyl** [reply]
I found that I can pass all the tests after several tests, and I have to repeat it several times each time all tests are successful

**Reply by godotalgorithm** [reply]
The input for some tests depend on the output of other tests. These inputs are statically generated, but they are also regenerated during the testing process. If this data is in the middle of being regenerated when the test runs, they will read in bad input data. CTest has a system to set the dependency of tests, but I didn't start using it until version 22.0.4. Once CTest knows the test dependencies, then something like `ctest -j16` works.

**Reply by 123xyl** [reply]
oh，got it

**Reply by 123xyl** [reply]
![image](https://user-images.githubusercontent.com/96686347/209440206-eb080e2d-95c9-473f-937f-84ecfc56438b.png)
![image](https://user-images.githubusercontent.com/96686347/209440213-6a8480a8-cde2-4327-8f12-9c3353fe170a.png)
What do these two scripts do?

**Reply by godotalgorithm** [reply]
That script will regenerate all output data that is used in the main set of MOPAC tests and delete some intermediate files that aren't used. I was using it a lot when I first wrote the CTest-based testing for MOPAC, but it isn't routinely used anymore.

The tests in MOPAC are all relative, not absolute. They test how MOPAC behaves now compared with how it behaved in the recent past. Because of this, I need to be careful with how this reference test data changes - it should only change when the behavior of MOPAC is intentionally being changed.

**Reply by 123xyl** [reply]
When I was in the build directory ctest for the first time, several tests failed. After I executed these two scripts, I went to the build directory ctest again, and all tests were passed

**Reply by godotalgorithm** [reply]
Except for the CTest concurrency bug, the main tests should always pass after you run that script because the test data and reference data will be generated by the same exact MOPAC executable. The reference data from the repository is a historical record of how MOPAC is supposed to run. If you regenerate the reference data, then you lose that history and the reference data just shows how MOPAC is running on your machine right now.
